library/std/src/sys/sync/condvar/queue.rs

```diff
@@ -0,0 +1,380 @@
+//! A generic `Condvar` implementation based on thread parking and a lockless
+//! queue of threads.
+//!
+//! Not all platforms provide an efficient `Condvar` implementation: the UNIX
+//! `pthread_condvar_t` needs memory allocation, while SGX doesn't have
+//! synchronization primitives at all. Therefore, we implement our own.
```

Comment:
Do we think this is the right tradeoff? It looks like the affected platforms are fairly "rare", and I'm not sure it makes sense for std to be maintaining relatively complex data structures for such platforms, especially if they're not even getting lightly tested in our CI.

Reply:
The problem is that every couple of months someone comes along and adds a completely new platform to `std`. Understandably, not every target maintainer is well-versed in writing synchronization primitives, leading to bugs like #114581 on SGX. Actually, SGX is an interesting case, because the environment *only* provides thread parking, so there'll always be the need for some queue like this one. The idea behind this PR is that this implementation will be used as the default on every new target, unless they have a good reason not to, so that we only have to maintain one admittedly complex implementation instead of five.

There are simpler implementations that I could switch to, if desired (the NetBSD `pthread_condvar` for example uses a stack of waiters and spins on contended access to it). It's just that making the shared implementation really good seemed like a good idea to me (it also might make it easier to convince people that they *really* shouldn't bring their own).

Reply:
I see. Yeah, if this is intended to be the fallback queue absent platform-specific, built-in queues that make more sense, then that makes more sense to me.